Tutorial (#282)
nf-core/nf-co.re: markdown/usage/nf_core_tutorial.md

In order to run _nf-core_ pipelines, you will need to have Nextflow installed ([https://www.nextflow.io](https://www.nextflow.io/)).
The only other requirement is a software packaging tool: [Conda](https://docs.conda.io/en/latest/miniconda.html), [Docker](https://www.docker.com) or [Singularity](https://sylabs.io/singularity/).
In theory it is possible to run the pipelines with software installed by other methods (_e.g._ environment modules, or manual installation), but this is not recommended.
Most people find either Docker or Singularity the best options.

### Fetching pipeline code

Unless you are actively developing pipeline code, we recommend using the Nextflow [built-in functionality](https://www.nextflow.io/docs/latest/sharing.html) to fetch _nf-core_ pipelines.
Nextflow will automatically fetch the pipeline code when you run `nextflow run nf-core/PIPELINE`.
For the best reproducibility, it is good to explicitly reference the pipeline version number that you wish to use with the `-revision`/`-r` flag.
For example:

```bash
nextflow run nf-core/rnaseq -revision 1.3
```

If not specified, Nextflow will fetch the default branch:

> * `master`  for _nf-core_ released pipelines, and will be the latest release

Reviewer: ```suggestion
```

> * `dev` for _nf-core_ pipelines still in development, and will contain the latest development code

Reviewer: ```suggestion
```

This launches an interactive prompt on the command line which asks for things such as pipeline name, a short description and the author's name.
These values are then propagated throughout the template files automatically.

### TODO statements

Not everything can be completed with a template and all new pipelines will need to edit and add to the resulting pipeline files in a similar set of locations.
To make it easier to find these, the _nf-core_ template files have numerous comment lines beginning with `TODO nf-core:`, followed by a description of what should be changed or added.
These comment lines can be deleted once the required change has been made.

Most code editors have tools to automatically discover such `TODO` lines and the `nf-core lint` command will flag these.
This makes it simple to systematically work through the new pipeline, editing all files where required.

### How nf-core software packaging works

The only hard requirement for all _nf-core_ pipelines is that software must be available in Docker images. However, it is recommended that pipelines use the following methodology where possible:

> 1. Software requirements are defined for Conda in `environment.yml`

Reviewer 2: Should we specify here that the packages should be available in the conda-forge / bioconda / etc. channels, as we had the case in the past that someone provided their own channel?

Reviewer: I guess it will still work if people use their own channel in the `environment.yml` file? I think that's probably a little too much detail for the tutorial though to be honest..

* PRs should be reviewed by someone else before being merged

> ### Setting up Docker and Travis CI

Author: That's the old one tutorial, I was thinking to remove it at the end, but at the moment, I kept it as it is.

When writing `conf/test.config` remember to define all required parameters so that the pipeline will run with only `-profile test`. Note that remote URLs cannot be traversed like a regular file system - so glob file expansions such as `*.fa` will not work.

> ### Travis CI configuration

Author: That's it :-D

* Take a look at `conf/test.config` and switch the test data for another dataset on nf-core/test_data.

## Releasing nf-core pipelines

Your pipeline is written and ready to go! Before you can release it with _nf-core_ there are a few steps that need to be done. First, tell everyone about it on Slack in the [`#new-pipelines`](https://nfcore.slack.com/channels/new-pipelines) channel. Hopefully you've already done this before you spent lots of time on your pipeline, to check that there aren't other similar efforts happening elsewhere. Next, you need to be a member of the [nf-core GitHub organisation](https://github.com/nf-core/). You can find instructions for how to do this at [https://nf-co.re/join](https://nf-co.re/join).

### Forking to nf-core

Once you're ready to go, you can fork your repository to _nf-core_. A lot of stuff happens automatically when you do this: the website will update itself to include your new pipeline, complete with rendered documentation pages and usage statistics. Your pipeline will also appear in the `nf-core list` command output and in various other locations.

> Unfortunately, at the time of writing, Travis CI, Docker Hub and Zenodo (automated DOI assignment for releases) services are not created automatically. These can only be set up by _nf-core_ administrators, so please ask someone to do this for you on Slack.

Author: I guess I should have explained myself more about that `nf_core_tutorial.md` file.
Sorry about that.

Typically, the pipeline should be able to complete in around 10 minutes and use no more than 6-7 GB memory.
To achieve this, input files and reference genomes need to be very tiny.
If possible, a good approach can be to use PhiX or Yeast as a reference genome.
Alternatively, a single small chromosome (or part of a chromosome) can be used.
If you are struggling to get the tests to run, ask for help on Slack.

When writing `conf/test.config` remember to define all required parameters so that the pipeline will run with only `-profile test`.
Note that remote URLs cannot be traversed like a regular file system - so glob file expansions such as `*.fa` will not work.

### GitHub Actions configuration

The automated tests with GitHub Actions are configured in the `.github/worfklows` folder that is generated by the template.
Each file Each file (`branch.yml`, `ci.yml` and `linting.yml`) defines several tests: branch protection, running the pipeline with the test data, linting the code with `nf-core lint`, linting the syntax of all Markdown documentation and linting the syntax of all YAML files.

#### `branch.yml`

> This is already set up and should not be touched.

Reviewer: ```suggestion
This is already set up and does not need to be edited.
```

> It will check that a PR is made on `dev` unless it's for a release or a patch.

Reviewer: ```suggestion
It will check that pull-requests going to the nf-core repo `master` branch only come from the `dev` or `patch` branches (for releases).
```

#### `linting.yml`

> This is already set up and should not be touched.

Reviewer: ```suggestion
This is already set up and does not need to be edited.
```

It will check:

* That all Markdown documentation follow a proper syntax.

> (if you're using `atom` of `Visual Studio Code` or any similar editor, plugins are available to help you out: _e.g._ [markdownlint for atom](https://atom.io/packages/linter-markdownlint), [markdownlint for Visual Studio Code](https://marketplace.visualstudio.com/items?itemName=DavidAnson.vscode-markdownlint))

Reviewer: ```suggestion
  * Many code editors have similar packages to help with markdown validation. For example, [markdownlint for atom](https://atom.io/packages/linter-markdownlint) and [markdownlint for Visual Studio Code](https://marketplace.visualstudio.com/items?itemName=DavidAnson.vscode-markdownlint))
  * You can run the markdown linting on the command line by [installing markdownlint](https://www.npmjs.com/package/markdownlint) and running the command `markdownlint . -c .github/markdownlint.yml`
```

> * That all YAML file follow a proper syntax.

Reviewer: Really? I can't find this in the template: https://github.com/nf-core/tools/blob/dev/nf_core/pipeline-template/%7B%7Bcookiecutter.name_noslash%7D%7D/.github/workflows/linting.yml

Author: https://github.com/nf-core/sarek/blob/5c30fd821ab7cdf27fd4105464a1fba5eabdfcd5/.github/workflows/linting.yml#L19-L31
forgot to make a PR for it...
I'll work on that

Reviewer: Nice 👍 

Note that the `|` character is only needed for multi-line strings.

So
```yaml
run: |
  yamllint $(find ${GITHUB_WORKSPACE} -type f -name "*.yml")
```

is probably nicer as just
```yaml
run: yamllint $(find ${GITHUB_WORKSPACE} -type f -name "*.yml")
```

Author: https://github.com/nf-core/tools/pull/534

> * That the code follow the nf-core linting.

Reviewer: ```suggestion
* That the pipeline code follows the nf-core linting rules.
  * You can run this manually with the `nf-core lint` command.
```

#### `ci.yml`

> 

Reviewer: ```suggestion

This file defines a GitHub Actions workflow that tests your pipeline with the small test dataset.
It should mostly work out of the box, but it may need some editing from you to customise it for your pipeline.

```

> The `nxf_ver` variable sets the `NXF_VER` environment variable twice.

Reviewer: ```suggestion
The matrix `nxf_ver` variable sets the `NXF_VER` environment variable twice.
```

This tells GitHub Actions to run the tests twice in parallel - once with the latest version of Nextflow (`NXF_VER=''`) and once with the minimum version supported by the pipeline.
Do not edit this version number manually - it appears in multiple locations through the pipeline code, so it's better to use `nf-core bump-version --nextflow` instead.

> The provided tests may be sufficient for your pipeline.

Reviewer: ```suggestion
The provided tests run your pipeline with the `-profile test,docker` flags. This may be sufficient for your pipeline.
```

However, if it is possible to run the pipeline with significantly different options (for example, different alignment tools), then it is good to test all of these.

> You can do this by adding additional tests in the `jobs` block.

Reviewer: ```suggestion
You can do this by adding additional tests in the `jobs` block.
Do not try add a run for every possible combination of parameters, as it will take too long to run.
```

This PR will never be merged, but gives the GitHub review web pages where people can comment on specific lines in the code.

These first community reviews can take quite a long time and typically result in a lot of comments and suggestions (nf-core/deepvariant famously had 156 comments before it was approved).
Try not to be intimidated - this is the main step where the community attempts to standardise and suggest improvements for your code.
Your pipeline will come out the other side stronger than ever!

### Making the first release

Once the pseudo-PR is approved, you're ready to make the release.
To do this, first bump the pipeline version to a stable tag using `nf-core bump-version`, then open a pull-request from the `dev` branch to `master`.
Once tests are passing and two _nf-core_ members have approved this PR, it can be merged to `master`.
Then a GitHub release is made, using the contents of the changelog as a description.

Pipeline version numbers (release tags) should be numerical only, using [semantic versioning](https://semver.org/spec/v2.0.0.html).

> For example, with a release version `1.4.3`, bumping `1` would correspond to the _major_ release where results would no longer be backwards compatible.

Reviewer: ```suggestion
For example, with a release version `1.4.3`, bumping to `2.0` would correspond to the _major_ release where results would no longer be backwards compatible.
```

> Changing `4` would be a minor release, for example adding some new features.

Reviewer: ```suggestion
Bumping to `1.5` would be a minor release, for example adding some new features.
```

> Changing `3` would be a patch release for minor things such as fixing bugs.

Reviewer: ```suggestion
Bumping to `1.4.4` would be a patch release for minor things such as fixing bugs.
```
